src/main/java/hudson/remoting/Engine.java

```diff
@@ -351,8 +342,15 @@ private Socket connect(String host, String port) throws IOException, Interrupted
                 }
                 return s;
             } catch (IOException e) {
-                if(retry++>10)
-                    throw (IOException)new IOException("Failed to connect to "+host+':'+port).initCause(e);
+            	if (retry++>10) {
+            		String suffix = "";
+            		if (isHttpProxy) {
+            			suffix = " through proxy " + targetAddress.toString();
+            		}
+            		// Cast -> InitCause method is returning a type Throwable.
+            		// Therefore we need to cast it to IOException.
+            		throw (IOException) new IOException("Failed to connect to " + host + ':' + port + suffix).initCause(e);
```

Comment:
NIT: I would decouple it to 3 lines and avoid any casts. BTW this approach should also work